src/KOKKOS/sna_kokkos_impl.h

```diff
@@ -532,79 +560,101 @@ KOKKOS_INLINE_FUNCTION
 void SNAKokkos<DeviceType>::compute_yi(int iter,
  const Kokkos::View<F_FLOAT**, DeviceType> &beta)
 {
-  double betaj;
-  const int iatom = iter / idxz_max;
-  const int jjz = iter % idxz_max;
-
-  const int j1 = idxz(jjz,0);
-  const int j2 = idxz(jjz,1);
-  const int j = idxz(jjz,2);
-  const int ma1min = idxz(jjz,3);
-  const int ma2max = idxz(jjz,4);
-  const int mb1min = idxz(jjz,5);
-  const int mb2max = idxz(jjz,6);
-  const int na = idxz(jjz,7);
-  const int nb = idxz(jjz,8);
-  const int jju = idxz(jjz,9);
-
-  const double* cgblock = cglist.data() + idxcg_block(j1,j2,j);
-  //int mb = (2 * (mb1min+mb2max) - j1 - j2 + j) / 2;
-  //int ma = (2 * (ma1min+ma2max) - j1 - j2 + j) / 2;
-
-  double ztmp_r = 0.0;
-  double ztmp_i = 0.0;
-
-  int jju1 = idxu_block[j1] + (j1+1)*mb1min;
-  int jju2 = idxu_block[j2] + (j2+1)*mb2max;
-  int icgb = mb1min*(j2+1) + mb2max;
-  for(int ib = 0; ib < nb; ib++) {
-
-    double suma1_r = 0.0;
-    double suma1_i = 0.0;
-
-    int ma1 = ma1min;
-    int ma2 = ma2max;
-    int icga = ma1min*(j2+1) + ma2max;
-
-    for(int ia = 0; ia < na; ia++) {
-      suma1_r += cgblock[icga] * (ulisttot(jju1+ma1,iatom).re * ulisttot(jju2+ma2,iatom).re - ulisttot(jju1+ma1,iatom).im * ulisttot(jju2+ma2,iatom).im);
-      suma1_i += cgblock[icga] * (ulisttot(jju1+ma1,iatom).re * ulisttot(jju2+ma2,iatom).im + ulisttot(jju1+ma1,iatom).im * ulisttot(jju2+ma2,iatom).re);
-      ma1++;
-      ma2--;
-      icga += j2;
-    } // end loop over ia
-
-    ztmp_r += cgblock[icgb] * suma1_r;
-    ztmp_i += cgblock[icgb] * suma1_i;
-    jju1 += j1+1;
-    jju2 -= j2+1;
-    icgb += j2;
-  } // end loop over ib
-
-  // apply to z(j1,j2,j,ma,mb) to unique element of y(j)
-  // find right y_list[jju] and beta(iatom,jjb) entries
-  // multiply and divide by j+1 factors
-  // account for multiplicity of 1, 2, or 3
-
-  // pick out right beta value
-
-  if (j >= j1) {
-    const int jjb = idxb_block(j1,j2,j);
-    if (j1 == j) {
-      if (j2 == j) betaj = 3*beta(jjb,iatom);
-      else betaj = 2*beta(jjb,iatom);
-    } else betaj = beta(jjb,iatom);
-  } else if (j >= j2) {
-    const int jjb = idxb_block(j,j2,j1);
-    if (j2 == j) betaj = 2*beta(jjb,iatom)*(j1+1)/(j+1.0);
-    else betaj = beta(jjb,iatom)*(j1+1)/(j+1.0);
-  } else {
-    const int jjb = idxb_block(j2,j,j1);
-    betaj = beta(jjb,iatom)*(j1+1)/(j+1.0);
-  }
+  int itriple;
+  for(int elem1 = 0; elem1 < nelements; elem1++)
+    for (int elem2 = 0; elem2 < nelements; elem2++) {
+      double betaj;
```

Comment:
You can move this line through the line reading `const double* cgblock = [...]` out of the loop over `elem1`, `elem2` to save extra loads for `nelements > 1`, since these are independent of the element.

Reply:
Same in `compute_zi`. It's trickier in `compute_bi` because of the hierarchical parallelism.

Reply:
Thanks @weinbe2 ! I just implemented this.

Reply:
@stanmoore1 I also updated a few other portions that will most likely increase performance.